docs/csharp/programming-guide/classes-and-structs/ref-returns.md

```diff
@@ -0,0 +1,94 @@
+---
+title: "Ref return values and ref locals (C# Guide) |  | Microsoft Docs"
+description: "Learn how to define and use ref return and ref local values"
+keywords: "ref returns, C#", "reference return values, C#", "ref return values, C#", "ref locals, C#", "ref local values, C#" 
+author: "rpetrusha"
+ms.author: "ronpet"
+ms.date: "05/10/2017"
+ms.topic: "article"
+ms.prod: ".net"
+ms.technology: "devlang-csharp"
+ms.devlang: "csharp"
+ms.assetid: "18cf7a4b-29f0-4b14-85b8-80af754aabd8"
+---
+# Ref returns and ref locals
+
+Starting with C# 7, C# supports reference return values (ref returns). A reference return value allows a method to return a reference to an object, rather than a value, back to a caller. The caller can then choose to treat the returned object returned as if it were returned by value or by reference. A value returned by reference that the caller handles as a reference rather than a value is a ref local).
+
+## What is a reference return value?
+
+Most developers are familiar with passing an argument to a called method *by reference*. A called method's argument list includes a value passed by reference, and any changes made to its value by the called method are returned to the caller. A *reference return value* is the opposite:
+
+- The called method's return value, rather than an argument passed to it, is a reference.
+
+- The caller, rather than the called method, can modify the value returned by the method.
+
+- Instead of modifications to the argument that are reflected in state of the object on the caller, modifications to the method's return value by the caller are reflected in the state of the object whose method was called.
+
+Reference return values can produce more compact code, as well as allow an object to expose only the individual data items, such as an array element, that are of interest to the caller. This reduces the likelihood that the caller will inadvertently modify the object's state.
+
+There are some restrictions on the value that a method can return as a reference return value. These include:
+
+- The return value cannot be `void`. Attempting to define a method with a `void` reference return value generates compiler error CS1547, "Keyword 'void' cannot be used in this context."
+ 
+- The return value cannot be a local variable in the procedure that returns it; it must have global scope. Attempting to return a local variable generates compiler error CS8168, "Cannot return local 'obj' by reference because it is not a ref local."
```

Comment:
> it must have global scope
That's not true. It must have a scope that is outside of the method: It can be a field of the object, or a static field of a class. It can also be an argument passed to the method.